packages/taiko-client/prover/proof_producer/combined_producer.go

```diff
@@ -17,24 +17,143 @@ import (
 	"github.com/taikoxyz/taiko-mono/packages/taiko-client/bindings/metadata"
 )
 
+type BlockProofState struct {
+	verifiedTiers []uint16
+	proofs        []encoding.SubProof
+}
+
+type ProofStateManager struct {
+	mu     sync.Mutex
+	states map[uint64]*BlockProofState
+}
+
+func NewProofStateManager() *ProofStateManager {
+	return &ProofStateManager{
+		states: make(map[uint64]*BlockProofState),
+	}
+}
+
+func (m *ProofStateManager) create(blockID *big.Int) {
+	blockIDUint64 := blockID.Uint64()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[blockIDUint64]
+	if !ok {
+		state = &BlockProofState{
+			verifiedTiers: []uint16{},
+			proofs:        []encoding.SubProof{},
+		}
+		m.states[blockIDUint64] = state
+	}
+}
+
+func (m *ProofStateManager) containsTier(blockID *big.Int, tier uint16) bool {
+	blockIDUint64 := blockID.Uint64()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[blockIDUint64]
+	if !ok {
+		return false
+	}
+
+	return slices.Contains(state.verifiedTiers, tier)
+}
+
+// addTierAndProof marks the tier as verified and adds the subproof to the block's state if
+// the state has not yet collected enough proofs. It returns whether the required number
+// of proofs has now been reached.
+func (m *ProofStateManager) addTierAndProof(
+	blockID *big.Int,
+	tier uint16,
+	subProof encoding.SubProof,
+	requiredProofs uint8,
+) (reachedRequired bool) {
+	blockIDUint64 := blockID.Uint64()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[blockIDUint64]
+	if !ok {
+		// create if it doesn't exist
+		state = &BlockProofState{
+			verifiedTiers: []uint16{},
+			proofs:        []encoding.SubProof{},
+		}
+		m.states[blockIDUint64] = state
+	}
+
+	// Record that we've verified this tier
```

Comment:
Same goes for this comment and 2 comments below.
IMO only stuff that isn't obvious from the code should be commented, and even then you should think could you named your values better, given semantic meaning by moving the code in separate function etc.

I only reach for comments if I can give valuable context to the reader that I think it's not easy to express through the code. Or, if I made any weird decisions to explain why.